feat: Monitor

adityahase opened the pull request:
Collect HTTP Request and Job logs

Collected data is put in `logs/monitor.json.log` in JSON format

e.g.

```
{"duration": 807142, "request": {"ip": "127.0.0.1",   ...  "request", "uuid": "83be6a4c-27a1-497a-9ce6-c815bca4e420"}
{"duration": 1364, "job": {"method": "frappe.ping",   ...   "uuid": "8225ab76-8bee-462c-b9fc-a556406b1ee7"}
{"duration": 26300, "job": {"method": "frappe.utils.  ...   "uuid": "e7c987e1-81ac-4d4b-845b-0fdd364aa963"}
```

Note: Set `monitor: true` in `site_config.json` or `common_site_config.json`

e.g. For a request
```JSON
{
    "duration": 807142,
    "request": {
        "ip": "127.0.0.1",
        "method": "GET",
        "path": "/api/method/frappe.realtime.get_user_info",
        "response_length": 9687,
        "status_code": 500
    },
    "site": "frappe.local",
    "timestamp": "2020-03-05 09:37:17.397884",
    "transaction_type": "request",
    "uuid": "83be6a4c-27a1-497a-9ce6-c815bca4e420"
} 
```
For an enqueued job
```JSON
{
    "duration": 1364,
    "job": {
        "method": "frappe.ping",
        "scheduled": false,
        "wait": 90204
    },
    "site": "frappe.local",
    "timestamp": "2020-03-05 09:37:40.124682",
    "transaction_type": "job",
    "uuid": "8225ab76-8bee-462c-b9fc-a556406b1ee7"
}
```
For a scheduled job
```JSON
{
    "duration": 26300,
    "job": {
        "method": "frappe.utils.global_search.sync_global_search",
        "scheduled": true,
        "wait": 10984295
    },
    "site": "frappe.local",
    "timestamp": "2020-03-05 09:38:03.158618",
    "transaction_type": "job",
    "uuid": "e7c987e1-81ac-4d4b-845b-0fdd364aa963"
}
```


Documentation PR: https://github.com/frappe/frappe_io/pull/261

adityahase committed:
feat: Monitor

Collect HTTP Request and Background Job logs

adityahase committed:
fix(monitor): Maintain MAX_LOG entries in cache

adityahase committed:
fix(monitor): Remove unused imports

adityahase committed:
fix(monitor): Add UUID to every log

adityahase committed:
fix(monitor): Use datetime.utcnow() instead of timezone.utc

Python 2, 3 compatibility issue

adityahase committed:
fix: Push a newline after all logs

adityahase committed:
fix(monitor): Include request ip in monitored data

surajshetty3416 commented on frappe/monitor.py:
```diff
@@ -0,0 +1,108 @@
+				"uuid": self.uuid,
+				"duration": self.duration,
+				"site": self.site,
+				"timestamp": self.timestamp.isoformat(sep=" "),
```
```suggestion
				"timestamp": self.timestamp.isoformat(sep=str(" ")),
```
to avoid...
<img width="1440" alt="Screenshot 2020-02-24 at 3 59 57 PM" src="https://user-images.githubusercontent.com/13928957/75214442-993e9780-57b3-11ea-94f2-dbfb96739a81.png">

surajshetty3416 reviewed (changes requested):
@adityahase when do you dump(flush) logs into a file?  have to do it manually?

adityahase commented:
> @adityahase when do you dump(flush) logs into a file? have to do it manually?

`frappe.monitor.flush` in `hooks.py`

adityahase committed:
refactor: Monitor

Do not collect request headers
Collect job wait time and whether it was scheduled

adityahase committed:
style: Remove unused imports

adityahase committed:
fix: Remove flaky assertions

pm-at commented:
When requests to just some specific methods need to be monitored, it would help to be able to specify filters for the same in config. This will minimize the load due to logging.
